va/validation-authority.go

```diff
@@ -96,17 +97,37 @@ func (va ValidationAuthorityImpl) validateSimpleHTTP(identifier core.AcmeIdentif
 	}
 	hostName := identifier.Value
 
+	addrs, _, _, err := va.DNSResolver.LookupHost(hostName)
+	if err != nil {
+		challenge.Status = core.StatusInvalid
+		setChallengeErrorFromDNSError(err, &challenge)
+		va.log.Debug(fmt.Sprintf("%s [%s] DNS failure: %s", challenge.Type, identifier, err))
+		return challenge, challenge.Error
+	}
+	if len(addrs) == 0 {
+		challenge.Error = &core.ProblemDetails{
+			Type:   core.UnknownHostProblem,
+			Detail: fmt.Sprintf("Could not resolve %s", hostName),
+		}
+		challenge.Status = core.StatusInvalid
+		return challenge, errors.New(challenge.Error.Detail)
+	}
+	challenge.ResolvedAddrs = addrs
+
 	var scheme string
 	if input.TLS == nil || (input.TLS != nil && *input.TLS) {
 		scheme = "https"
 	} else {
 		scheme = "http"
 	}
+	var dvsniAddr string
 	if va.TestMode {
-		hostName = "localhost:5001"
+		dvsniAddr = "localhost:5001"
+	} else {
+		dvsniAddr = addrs[0].String()
 	}
 
-	url := fmt.Sprintf("%s://%s/.well-known/acme-challenge/%s", scheme, hostName, challenge.Path)
+	url := fmt.Sprintf("%s://%s/.well-known/acme-challenge/%s", scheme, dvsniAddr, challenge.Path)
```

Comment:
Please also use the net.URL struct to generate these. There'll be boogs that show up in weird places w.r.t. escaping.